Add Fishing Event System for Autofisher ModCall Integration

hocha113 opened the pull request:
## 变更概述

为自动钓鱼机添加了完整的钓鱼事件系统，允许外部模组通过 ModCall 注册钓鱼事件，在钓鱼机钓起物品前对钓鱼结果进行修改或取消

## 动机和背景

### 问题

当前自动钓鱼机缺乏让外部模组干预钓鱼过程的机制，导致外部模组难以做到以下需求：
•	无法实现钓鱼结果的动态修改
•	无法添加自定义钓鱼逻辑（如特殊环境下的额外奖励）
•	无法过滤或替换特定物品

### 对此这次PR所给出的解决方案

新增钓鱼事件ModCall系统，为外部模组提供以下能力：
•	事件注册：外部模组可注册回调函数，在钓鱼机收杆前执行
•	结果修改：可修改钓上的物品类型和数量
•	事件取消：可阻止本次钓鱼成功

#### 使用示例

委托实例，签名为 `void Method(TileEntity fisher, FishingAttempt fishingAttempt, Player player, ref int itemType, ref int itemStack, ref bool cancel)`

#### 使用示例

```csharp
public class MyMod : Mod
{
    public override void PostSetupContent()
    {
        if (!ModLoader.TryGetMod("ImproveGame", out Mod improveGame))
                return;

        // 定义委托签名并转换
        improveGame.Call("RegisterFishingEvent", (Delegate)OnFishingCallback);
    }
    
    private void OnFishingCallback(TileEntity fisher, FishingAttempt fishingAttempt, Player player, ref int itemType, ref int itemStack, ref bool cancel)//注意这里的签名需要和委托定义一致
    {
        // 直接通过 ref 参数修改
        itemType = ItemID.IronCrate;//全部变成铁匣子
    }

    //这里可以偷懒不卸载委托，因为模组在重新加载时本身会自动卸载清理所有的委托
}
```
### 感谢审查！🙏

hocha113 committed:
Add fishing event system for mod extensibility

Introduce a flexible fishing event system allowing other mods to modify fishing machine results or cancel fishing actions via registered callbacks. This enhances cross-mod integration and provides greater extensibility for custom fishing behaviors.

hocha113 committed:
Add weak reference support for fishing events

Enhanced mod interoperability by introducing weak reference-based event handlers for fishing events, allowing external mods to interact with reduced coupling. Retained strong reference handlers for compatibility. Improved code clarity with detailed comments.

hocha113 committed:
Update ModIntegrationsSystem.cs

hocha113 committed:
Merge branch 'ForOne-Club:1.4.4' into 1.4.4

hocha113 committed:
Add flexible fishing event callback support

Introduces a new multi-parameter delegate for fishing event callbacks, enabling more flexible and efficient event handling. Updated event argument structure for better performance and added support for multiple registration methods, including strong and weak references. Enhanced documentation with examples and improved event invocation logic for clarity and robustness.

hocha113 committed:
Update CrossModSupport.md

hocha113 committed:
Update CrossModSupport.md

487666123 commented:
## 有一些问题也许你可以处理下：

### 可以将 List 保存事件的更换为 event，event 的事件列表也可以获取。

```csharp
internal static event EventHandler<float> Foo;

public static void Fooo()
{
    foreach (EventHandler<float> e in Foo.GetInvocationList())
    {
        // ...
    }
}
```

### Mod.Call

Mod.Call 本身就是为弱引用设计的，强引用不需要 Mod.Call 方法，所以只需要设计一个弱引用版本的方法的即可。一个 Call 也不需要多个不同的版本，一个通用的方法即可。

### 不太重要的一点

通常也不需要设计一个新的 delegate，C# 有设计一个专用于此类的 delegate

```csharp
namespace System
{
    //
    // 摘要:
    //     Represents the method that will handle an event when the event provides data.
    //
    //
    // 参数:
    //   sender:
    //     The source of the event.
    //
    //   e:
    //     An object that contains the event data.
    //
    // 类型参数:
    //   TEventArgs:
    //     The type of the event data generated by the event.
    public delegate void EventHandler<TEventArgs>(object? sender, TEventArgs e);
}
```

事件参数通常如下方法设计：

```csharp
public class FooEventArgs : EventArgs
{
    public int Id { get; set; }
}
```

使用：

```csharp
public static event EventHandler<FooEventArgs> Foo;
```

487666123 commented:
事件参数的设计最好也不要使用可访问的字段，转而使用属性。
强引用可以直接订阅事件，所以直接设为 public 即可。

hocha113 committed:
Refactor fishing event system for modularity

Replaced the old fishing event system with a new, modular design centralized in the TEAutofisher class. This improves maintainability by consolidating event logic within the relevant entity, simplifies event registration and unregistration, and enhances flexibility for external mod integrations. Updated documentation to reflect the new event system and removed outdated examples.

hocha113 commented:
> 事件参数的设计最好也不要使用可访问的字段，转而使用属性。 强引用可以直接订阅事件，所以直接设为 public 即可。

受教了🙏，我进行了一些调整，首先是去除了 `List<Delegate> ` 的使用转而使用更加合规的 `event` ，并将其定义移动到 `TEAutofisher` 中。另外两种参数版本的Call注册我也删除了，只保留了
 `void Method(TileEntity fisher, FishingAttempt fishingAttempt, Player player, ref int itemType, ref int itemStack, ref bool cancel)` 
这个注册版本。
因为我想尽力避免反射的使用，所以我保留了
 `Delegate.CreateDelegate` 和 `FishingEventCallback` 
这样外部模组如果使用Call的话就可以让参数暴露出来，不必反射
